Kuadrant/kcp-glbc
# TLS support

This PR adds TLS support to the GLBC.

To provide this we setup cert-manager in a local kind cluster (make local setup will setup this cluster). This cluster is known as the control cluster. We then provide the k8sconfig to GLBC to allow it to interact with the local control cluster. When not run locally this control cluster would be the same cluster as GLBC is deployed in.

During start up GLBC will bootstrap a cert issuer based on flags passed in. A follow up here may be to add a CRD to control this rather than flags.

The basic flow it follows is :

- root ingress created
- tls enabled so ingress is given a tls section under its spec
- after DNS , certificate resource created in control cluster (control cluster locally is a kind cluster. It represent a cluster where glbc would be deployed) (dns verification performed by cert-manager)
- secrets controller watches for new secrets in the control cluster
- when a new tls secret is created secrets controller mirrors that secret into the correct workspace and namespace
- ingress controller notified tls is read, creates the leaf ingress objects and copies the root secret for each leaf ingress and adds it to the leaf TLS section. Each copied secret is labelled with a cluster to sync to
- Leaf ingress and secrets are synced to the correct pcluster


**Verification**

This can be verified locally or using a remote cluster.

Steps:

- you will need to set local env vars. To do this I create a .env file with the following values
```
export AWS_DNS_PUBLIC_ZONE_ID=ID
export AWS_ACCESS_KEY_ID=ID
export AWS_SECRET_ACCESS_KEY=TOPSECRET
export HCG_LE_EMAIL=YOUR_EMAIL

```

- ```make local-setup```
- run the ingress controller with the tls flags ```./bin/ingress-controller -kubeconfig .kcp/admin.kubeconfig -glbc-kubeconfig ./tmp/kcp-cluster-glbc-control.kubeconfig -glbc-tls-provided=true ```
- create the echo service
- create the ingress object 
- once the certificate has been issued the leaf ingresses will be created ```kubectl get ingress -n <namespace>```
- verify the secrets are also there ``` kubectl get secrets -n <namespace>
